keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1075 @@
+```
+
+### CRI API
+
+Add `bool recursive_read_only` to the [`Mount`](https://github.com/kubernetes/cri-api/blob/v0.26.1/pkg/apis/runtime/v1/api.proto#L212-L224) message.
+CRI implementations will also expose the availability of the feature via the `RuntimeHandlerFeatures` message.
+
+As kubelet can inspect the availability of the feature via the `RuntimeHandlerFeatures` message,
+there is no concept of "IfPossible" in the CRI API;
+kubelet translates an "IfPossible" value in the Core API into true or false in the CRI API
+
+The `RuntimeHandlerFeatures` message is also propagated to the `NodeSystemInfo` struct of the Core API.
+
+Diff:
+```diff
+diff --git a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+index 11d58636369..5852841a5bd 100644
+--- a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
++++ b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+@@ -226,6 +226,15 @@ message Mount {
+     repeated IDMapping uidMappings = 6;
+     // GidMappings specifies the runtime GID mappings for the mount.
+     repeated IDMapping gidMappings = 7;
++    // If set to true, the mount is made recursive read-only.
++    // In this CRI API, recursive_read_only is a plain true/false boolean, although its equivalent
++    // in the Kubernetes core API is a quaternary that can be nil, "Enabled", "IfPossible", or "Disabled".
++    // kubelet translates that quaternary value in the core API into a boolean in this CRI API.
++    // Remarks:
++    // - nil is just treated as false
++    // - when set to true, readonly must be explicitly set to true, and propagation must be PRIVATE (0).
++    // - (readonly == false && recursive_read_only == false) does not make the mount read-only.
++    bool recursive_read_only = 8;
+ }
+ 
+ // IDMapping describes host to container ID mappings for a pod sandbox.
+@@ -1512,6 +1521,15 @@ message StatusRequest {
+     bool verbose = 1;
+ }
+ 
++// RuntimeHandlerFeatures describes the features that are supported by the runtime handler.
++message RuntimeHandlerFeatures {
++    // recursive_read_only_mounts is set to true if the runtime handler supports
++    // recursive read-only mounts.
++    // For runc-compatible runtimes, availability of this feature can be detected by checking whether
++    // the Linux kernel version is >= 5.12, and,  `runc features | jq .mountOptions` contains "rro".
++    bool recursive_read_only_mounts = 1;
++}
++
+ message StatusResponse {
+     // Status of the Runtime.
+     RuntimeStatus status = 1;
+@@ -1520,6 +1538,8 @@ message StatusResponse {
+     // debug, e.g. plugins used by the container runtime.
+     // It should only be returned non-empty when Verbose is true.
+     map<string, string> info = 2;
++    // key = runtime handler name. The empty string denotes the default handler.
++    map<string, RuntimeHandlerFeatures> features = 3;
```

Comment:
We have #4033 and #3004 doing roughly similar thing so we need to coordinate with this. @mrunalp

Reply:
Yes, agree. For that we had discussed that it would make sense to have a separate CRI call for feature discovery.

Reply:
> should we add a streaming interface rather than yet another continuosly polled endpoint? I have the future with more dynamically changing config/features in mind here.

Sounds overengineering.
The runtime classes rarely changes through the lifecycle of a node, and I guess we can just expect the admin to send `SIGHUP` to the CRI runtime and kubelet.

> Also, what do you think about a list with named fields instead of a map?

Pros and cons?

Reply:
> Sounds overengineering. 

For this particular use case, yes. But I was thinking the future extensions for dynamic configuration/features. I think we have at least these alternatives:

1. Extend RuntimeStatus as you proposed. Future extensions will follow this and it works in practice as the runtime status is being continuously polled by the kubelet
2. Add a new API endpoint, say RuntimeConfig (what @mrunalp) suggested. With the future extensions we start to need polling this too. Now we have two endpoints being continuously polled (RuntimeStatus and RuntimeConfig) which sounds silly to me (I think it would be a lot cheaper to just poll RuntimeStatus with a little bit more data in it).
3. Have 2 above for "static'ish" information that only gets invoked at kubelet startup and runtime restart or smth. In addition to that, add a streaming API for the dynamic use case. Now we have two endpoints for the same information 🤔
4. Have just the streaming API for all usage scenarios

Maybe it's just me but 2. and 3. don't sound attractive from the CRI API pov 

> Pros and cons?

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
> For this particular use case, yes. But I was thinking the future extensions for dynamic configuration/features. I think we have at least these alternatives:

Yes, but I think this is beyond the scope of this KEP.

> https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Doesn't seem relevant to CRI API.

Reply:
> > https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
> 
> Doesn't seem relevant to CRI API.

But I think it's the prevalent convention. I'm only after consistency here. (I was also instructed to forget about maps, see https://github.com/kubernetes/enhancements/pull/3004#discussion_r1099644184)

Reply:
Removed the map